README.md
# @mikhail-g/node-resemble-js

Analyse and compare images with Javascript. This project does not need canvas or any other binary dependencies.
It is a modification of [Resemble.js](https://github.com/Huddle/Resemble.js)

## Installation

`npm install @mikhail-g/node-resemble-js`

## Example

### Retrieve basic analysis on image

```javascript
const resemble = require('@mikhail-g/node-resemble-js');

resemble(fileData)
  .onComplete(function(data){
    console.log(data);
  });
```

output for a blank white image:

```javascript
 {
   red: 255,
   green: 255,
   blue: 255,
   brightness: 255
 }
```

### Use resemble to compare two images

```javascript
var diff = resemble(file)
  .compareTo(file2)
  .ignoreColors()
  .onComplete(function(data){
    console.log(data);
});
```

output:

```javascript
 {
   misMatchPercentage : 100, // %, when all pixels are different in 2 images
   isSameDimensions: false, // is width and height of 2 images are the same
   dimensionDifference: { width: 0, height: -1 }, // defined if dimensions are not the same
   getImageDataUrl: function(){}
 }
```

### Configuration

You can change the comparison method after the first analysis or before `onComplete()`.

```javascript
diff.ignoreNothing(); // any change in pixel color or brightness will increase misMatchPercentage
diff.ignoreColors(); // compares only brightness of pixels ignoring colors
diff.ignoreAntialiasing(); 
```

And change the output display style.

```javascript
resemble.outputSettings({
  errorColor: {
    red: 255,
    green: 0,
    blue: 255
  },
  errorType: 'movement',
  transparency: 0.3
});
// resembleControl.repaint();
```

--------------------------------------

### Usuage in cucumber step definition

``` javascript
     this.Then(/^Screenshot should match image "(.*)"$/, function (image, callback) {
        browser.takeScreenshot().then(function(pngString) {
          var screenshot = new Buffer(pngString, 'base64');
     
          resemble(image)
            .compareTo(screenshot)
            .onComplete(function(data){
     
              if (Number(data.misMatchPercentage) <= 0.01) {
                callback();
              } else {
                data.getDiffImage().pack().pipe(fs.createWriteStream(image + 'diff.png'));
                callback.fail(new Error("Screenshot '" + image+  "' differ " + data.misMatchPercentage + "%"));
              }
            });
        });
      })
```

--------------------------------------

## Credits

* Created by [James Cryer](https://github.com/jamescryer) and the Huddle development team.
* [Lukas Svoboda](https://github.com/lksv) - modification for node.js
* [Mirza Zeyrek](https://github.com/mirzazeyrek) - jpeg support
* [LarryG](https://github.com/larryg01) - dependency updates
* [Mykhailo Hariachyi](https://github.com/mikhail-g) - fix for ignoreNothing() function


CHANGELOG.md
# Changelog

## 1.3.0

* fixed `ingnoreNothing()` function to have 0 tolerance
* rename package name to `node-resemble-js`
* ignored `resemble.spec.js` and `example` images folder for npm release
* added licence notice for the forked repo
* changed author to include myself
* fixed and updated `README.md`
* updated `jpeg-js` to `"^0.4.4"`
* changed from gulp to jest

## 1.2.0

* updated jpeg-js and pngjs dependencies
* updated credits README.md

## 0.2.0

* added support for jpeg

## 0.1.0

* wrong release - kept back

## 0.0.4

* rename package name to node-resemble-js

## 0.0.3

* added ignoreRectangles feature

## 0.0.0

* initial commit (started form resemble sha 5ce379b)
* rewrite resence to node.js (with only pngjs dependency)
* added example


resemble.spec.js
'use strict';

describe('node-resemble.js', function() {
  var EXAMPLE_LARGE_IMAGE = 'example/LargeImage.png';
  var EXAMPLE_SMALL_IMAGE = 'example/SmallImage.png';
  var EXAMPLE_PEOPLE_IMAGES = [
	  'example/People.png',
	  'example/People2.png'
  ];
  var EXAMPLE_TOLERANCE_IMAGES = [
    'example/100x100_rgb_220.png',
    'example/100x100_rgb_220_with_change.png',
  ];

  var OPTIMISATION_SKIP_STEP = 6;
  var DEFAULT_LARGE_IMAGE_THRESHOLD = 1200;

  var resemble = require('./resemble.js');

  describe('largeImageThreshold', function() {
    describe('when unset', function() {
      describe('when ignoreAntialiasing is enabled', function() {
        it('skips pixels when a dimension is larger than the default threshold (1200)', function(done) {
          getLargeImageComparison().ignoreAntialiasing().onComplete(function(data) {
            expectPixelsToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        it('does not skip pixels when both dimensions are smaller than the default threshold (1200)', function(done) {
          getSmallImageComparison().ignoreAntialiasing().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });
      });

      describe('when ignoreAntialiasing is disabled', function() {
        it('does not skip pixels when a dimension is larger than the default threshold (1200)', function(done) {
          getLargeImageComparison().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        it('does not skip pixels when both dimensions are smaller than the default threshold (1200)', function(done) {
          getSmallImageComparison().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });
      });
    });

    describe('when explicitly set', function() {
      afterAll(function() {
        resemble.outputSettings({largeImageThreshold: DEFAULT_LARGE_IMAGE_THRESHOLD});
      });

      describe('when ignoreAntialiasing is enabled', function() {
        it('skips pixels on images with a dimension larger than the given threshold', function(done) {
          resemble.outputSettings({largeImageThreshold: 999});
          getSmallImageComparison().ignoreAntialiasing().onComplete(function(data) {
            expectPixelsToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        it('does not skip pixels on images with a dimension equal to the given threshold', function(done) {
          resemble.outputSettings({largeImageThreshold: 1000});
          getSmallImageComparison().ignoreAntialiasing().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        it('does not skip pixels on images with both dimensions smaller than the given threshold', function(done) {
          resemble.outputSettings({largeImageThreshold: 1001});
          getSmallImageComparison().ignoreAntialiasing().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });
      });

      describe('when ignoreAntialiasing is disabled', function() {
        it('does not skip pixels on images with a dimension larger than the given threshold', function(done) {
          resemble.outputSettings({largeImageThreshold: 999});
          getSmallImageComparison().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        it('does not skip pixels on images with a dimension equal to the given threshold', function(done) {
          resemble.outputSettings({largeImageThreshold: 1000});
          getSmallImageComparison().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        it('does not skip pixels on images with both dimensions smaller than the given threshold', function(done) {
          resemble.outputSettings({largeImageThreshold: 1001});
          getSmallImageComparison().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });
      });
    });

    describe('when set to a falsy value', function() {
      beforeEach(function() {
        resemble.outputSettings({largeImageThreshold: 0});
      });

      afterAll(function() {
        resemble.outputSettings({largeImageThreshold: DEFAULT_LARGE_IMAGE_THRESHOLD});
      });

      describe('when ignoreAntialiasing is enabled', function() {
        it('does not skip pixels on images with a dimension larger than the default threshold (1200)', function(done) {
          getLargeImageComparison().ignoreAntialiasing().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        it('does not skip pixels on images with a dimension smaller than the default threshold (1200)', function(done) {
          getSmallImageComparison().ignoreAntialiasing().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });
      });

      describe('when ignoreAntialiasing is disabled', function() {
        it('does not skip pixels on images with a dimension larger than the default threshold (1200)', function(done) {
          getLargeImageComparison().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        it('does not skip pixels on images with a dimension smaller than the default threshold (1200)', function(done) {
          getSmallImageComparison().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });
      });
    });

    function expectPixelsToBeSkipped(image, step) {
      expect(getPixelForLocation(image, 1, step - 1).alpha).not.toBe(0);
      expect(getPixelForLocation(image, 1, step).alpha).toBe(0);
      expect(getPixelForLocation(image, 1, step + 1).alpha).not.toBe(0);

      expect(getPixelForLocation(image, step - 1, 1).alpha).not.toBe(0);
      expect(getPixelForLocation(image, step, 1).alpha).toBe(0);
      expect(getPixelForLocation(image, step + 1, 1).alpha).not.toBe(0);

      expect(getPixelForLocation(image, step, step).alpha).toBe(0);
    }

    function expectPixelsNotToBeSkipped(image, step) {
      expect(getPixelForLocation(image, 1, step).alpha).not.toBe(0);
      expect(getPixelForLocation(image, step, 1).alpha).not.toBe(0);
      expect(getPixelForLocation(image, step, step).alpha).not.toBe(0);
    }
  });

  it('rawMisMatchPercentage contains raw result', function(done) {
	  resemble(
	  	EXAMPLE_PEOPLE_IMAGES[0]
	  ).compareTo(
	  	EXAMPLE_PEOPLE_IMAGES[1]
	  ).onComplete(function(data) {
		  expect(data.rawMisMatchPercentage).toBe(8.6612);
		  done();
	  });
  });

  describe('Minor color diff tolerance', function () {
    it('is applied by default', function (done) {
      resemble(EXAMPLE_TOLERANCE_IMAGES[0])
        .compareTo(EXAMPLE_TOLERANCE_IMAGES[1])
        .onComplete(function (data) {
          expect(data.misMatchPercentage).toBe('0.00');
          done();
        });
    });

    it('is not applied when ignoreNothing is used', function (done) {
      resemble(EXAMPLE_TOLERANCE_IMAGES[0])
        .compareTo(EXAMPLE_TOLERANCE_IMAGES[1])
        .ignoreNothing()
        .onComplete(function (data) {
          expect(data.misMatchPercentage).toBe('83.71');
          done();
        });
    });
  });

  function getLargeImageComparison() {
    return resemble(EXAMPLE_LARGE_IMAGE).compareTo(EXAMPLE_LARGE_IMAGE);
  }

  function getSmallImageComparison() {
    return resemble(EXAMPLE_SMALL_IMAGE).compareTo(EXAMPLE_SMALL_IMAGE);
  }

  function getPixelForLocation(image, x, y) {
    var index = (image.width * y + x) << 2;
    return {
      red: image.data[index],
      green: image.data[index + 1],
      blue: image.data[index + 2],
      alpha: image.data[index + 3]
    };
  }
});


resemble.test.js
'use strict';

describe('node-resemble.js', function() {
  var EXAMPLE_LARGE_IMAGE = 'example/LargeImage.png';
  var EXAMPLE_SMALL_IMAGE = 'example/SmallImage.png';
  var EXAMPLE_PEOPLE_IMAGES = [
	  'example/People.png',
	  'example/People2.png'
  ];
  var EXAMPLE_TOLERANCE_IMAGES = [
    'example/100x100_rgb_220.png',
    'example/100x100_rgb_220_with_change.png',
  ];

  var OPTIMISATION_SKIP_STEP = 6;
  var DEFAULT_LARGE_IMAGE_THRESHOLD = 1200;

  var resemble = require('./resemble.js');

  describe('largeImageThreshold', function() {
    describe('when unset', function() {
      describe('when ignoreAntialiasing is enabled', function() {
        test('skips pixels when a dimension is larger than the default threshold (1200)', function(done) {
          getLargeImageComparison().ignoreAntialiasing().onComplete(function(data) {
            expectPixelsToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        test('does not skip pixels when both dimensions are smaller than the default threshold (1200)', function(done) {
          getSmallImageComparison().ignoreAntialiasing().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });
      });

      describe('when ignoreAntialiasing is disabled', function() {
        test('does not skip pixels when a dimension is larger than the default threshold (1200)', function(done) {
          getLargeImageComparison().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        test('does not skip pixels when both dimensions are smaller than the default threshold (1200)', function(done) {
          getSmallImageComparison().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });
      });
    });

    describe('when explicitly set', function() {
      afterAll(function() {
        resemble.outputSettings({largeImageThreshold: DEFAULT_LARGE_IMAGE_THRESHOLD});
      });

      describe('when ignoreAntialiasing is enabled', function() {
        test('skips pixels on images with a dimension larger than the given threshold', function(done) {
          resemble.outputSettings({largeImageThreshold: 999});
          getSmallImageComparison().ignoreAntialiasing().onComplete(function(data) {
            expectPixelsToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        test('does not skip pixels on images with a dimension equal to the given threshold', function(done) {
          resemble.outputSettings({largeImageThreshold: 1000});
          getSmallImageComparison().ignoreAntialiasing().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        test('does not skip pixels on images with both dimensions smaller than the given threshold', function(done) {
          resemble.outputSettings({largeImageThreshold: 1001});
          getSmallImageComparison().ignoreAntialiasing().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });
      });

      describe('when ignoreAntialiasing is disabled', function() {
        test('does not skip pixels on images with a dimension larger than the given threshold', function(done) {
          resemble.outputSettings({largeImageThreshold: 999});
          getSmallImageComparison().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        test('does not skip pixels on images with a dimension equal to the given threshold', function(done) {
          resemble.outputSettings({largeImageThreshold: 1000});
          getSmallImageComparison().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        test('does not skip pixels on images with both dimensions smaller than the given threshold', function(done) {
          resemble.outputSettings({largeImageThreshold: 1001});
          getSmallImageComparison().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });
      });
    });

    describe('when set to a falsy value', function() {
      beforeEach(function() {
        resemble.outputSettings({largeImageThreshold: 0});
      });

      afterAll(function() {
        resemble.outputSettings({largeImageThreshold: DEFAULT_LARGE_IMAGE_THRESHOLD});
      });

      describe('when ignoreAntialiasing is enabled', function() {
        test('does not skip pixels on images with a dimension larger than the default threshold (1200)', function(done) {
          getLargeImageComparison().ignoreAntialiasing().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        test('does not skip pixels on images with a dimension smaller than the default threshold (1200)', function(done) {
          getSmallImageComparison().ignoreAntialiasing().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });
      });

      describe('when ignoreAntialiasing is disabled', function() {
        test('does not skip pixels on images with a dimension larger than the default threshold (1200)', function(done) {
          getLargeImageComparison().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });

        test('does not skip pixels on images with a dimension smaller than the default threshold (1200)', function(done) {
          getSmallImageComparison().onComplete(function(data) {
            expectPixelsNotToBeSkipped(data.getDiffImage(), OPTIMISATION_SKIP_STEP);
            done();
          });
        });
      });
    });

    function expectPixelsToBeSkipped(image, step) {
      expect(getPixelForLocation(image, 1, step - 1).alpha).not.toBe(0);
      expect(getPixelForLocation(image, 1, step).alpha).toBe(0);
      expect(getPixelForLocation(image, 1, step + 1).alpha).not.toBe(0);

      expect(getPixelForLocation(image, step - 1, 1).alpha).not.toBe(0);
      expect(getPixelForLocation(image, step, 1).alpha).toBe(0);
      expect(getPixelForLocation(image, step + 1, 1).alpha).not.toBe(0);

      expect(getPixelForLocation(image, step, step).alpha).toBe(0);
    }

    function expectPixelsNotToBeSkipped(image, step) {
      expect(getPixelForLocation(image, 1, step).alpha).not.toBe(0);
      expect(getPixelForLocation(image, step, 1).alpha).not.toBe(0);
      expect(getPixelForLocation(image, step, step).alpha).not.toBe(0);
    }
  });

  test('rawMisMatchPercentage contains raw result', function(done) {
	  resemble(
	  	EXAMPLE_PEOPLE_IMAGES[0]
	  ).compareTo(
	  	EXAMPLE_PEOPLE_IMAGES[1]
	  ).onComplete(function(data) {
		  expect(data.rawMisMatchPercentage).toBe(8.6612);
		  done();
	  });
  });

  describe('Minor color diff tolerance', function () {
    test('is applied by default', function (done) {
      resemble(EXAMPLE_TOLERANCE_IMAGES[0])
        .compareTo(EXAMPLE_TOLERANCE_IMAGES[1])
        .onComplete(function (data) {
          expect(data.misMatchPercentage).toBe('0.00');
          done();
        });
    });

    test('is not applied when ignoreNothing is used', function (done) {
      resemble(EXAMPLE_TOLERANCE_IMAGES[0])
        .compareTo(EXAMPLE_TOLERANCE_IMAGES[1])
        .ignoreNothing()
        .onComplete(function (data) {
          expect(data.misMatchPercentage).toBe('83.71');
          done();
        });
    });
  });

  function getLargeImageComparison() {
    return resemble(EXAMPLE_LARGE_IMAGE).compareTo(EXAMPLE_LARGE_IMAGE);
  }

  function getSmallImageComparison() {
    return resemble(EXAMPLE_SMALL_IMAGE).compareTo(EXAMPLE_SMALL_IMAGE);
  }

  function getPixelForLocation(image, x, y) {
    var index = (image.width * y + x) << 2;
    return {
      red: image.data[index],
      green: image.data[index + 1],
      blue: image.data[index + 2],
      alpha: image.data[index + 3]
    };
  }
});


resemble.js
/*
James Cryer / Huddle 2014
URL: https://github.com/Huddle/Resemble.js
*/
'use strict';

const PNG = require('pngjs').PNG;
const fs = require('fs');
const jpeg = require('jpeg-js');

//keeping wrong indentation and '_this' for better diff with origin resemble.js
let _this = {};

	let pixelTransparency = 1;

	let errorPixelColor = { // Color for Error Pixels. Between 0 and 255.
		red: 255,
		green: 0,
		blue: 255,
		alpha: 255
	};

	let errorPixelTransform = {
		flat : function (d1, d2){
			return {
				r: errorPixelColor.red,
				g: errorPixelColor.green,
				b: errorPixelColor.blue,
				a: errorPixelColor.alpha
			}
		},
		movement: function (d1, d2){
			return {
				r: ((d2.r*(errorPixelColor.red/255)) + errorPixelColor.red)/2,
				g: ((d2.g*(errorPixelColor.green/255)) + errorPixelColor.green)/2,
				b: ((d2.b*(errorPixelColor.blue/255)) + errorPixelColor.blue)/2,
				a: d2.a
			}
		}
	};

	let errorPixelTransformer = errorPixelTransform.flat;
	let largeImageThreshold = 1200;

	_this['resemble'] = function( fileData ){

		let data = {};
		let images = [];
		let updateCallbackArray = [];

		let tolerance = { // between 0 and 255
			red: 16,
			green: 16,
			blue: 16,
			alpha: 16,
			minBrightness: 16,
			maxBrightness: 240
		};

		let ignoreAntialiasing = false;
		let ignoreColors = false;
    let ignoreRectangles = null;

		function triggerDataUpdate(){
			let len = updateCallbackArray.length;
			let i;
			for(i=0;i<len;i++){
				if (typeof updateCallbackArray[i] === 'function'){
					updateCallbackArray[i](data);
				}
			}
		}

		function loop(x, y, callback){
			let i,j;

			for (i=0;i<x;i++){
				for (j=0;j<y;j++){
					callback(i, j);
				}
			}
		}

		function parseImage(sourceImageData, width, height){

			let pixleCount = 0;
			let redTotal = 0;
			let greenTotal = 0;
			let blueTotal = 0;
			let brightnessTotal = 0;

			loop(height, width, function(verticalPos, horizontalPos){
				let offset = (verticalPos*width + horizontalPos) * 4;
				let red = sourceImageData[offset];
				let green = sourceImageData[offset + 1];
				let blue = sourceImageData[offset + 2];
				let brightness = getBrightness(red,green,blue);

				pixleCount++;

				redTotal += red / 255 * 100;
				greenTotal += green / 255 * 100;
				blueTotal += blue / 255 * 100;
				brightnessTotal += brightness / 255 * 100;
			});

			data.red = Math.floor(redTotal / pixleCount);
			data.green = Math.floor(greenTotal / pixleCount);
			data.blue = Math.floor(blueTotal / pixleCount);
			data.brightness = Math.floor(brightnessTotal / pixleCount);

			triggerDataUpdate();
		}

		function loadImageData( fileData, callback ){

			if (Buffer.isBuffer(fileData)) {
				let png = new PNG();
				png.parse(fileData, function (err, data) {
				  callback(data, data.width, data.height);
				});
			} else {
				let ext = fileData.substring(fileData.lastIndexOf(".")+1);
				if(ext=="png") {
					let png = new PNG();
					fs.createReadStream(fileData)
					  .pipe(png)
					  .on('parsed', function() {
						callback(this, this.width, this.height);
					  });
				}
				if(ext=="jpg" || ext=="jpeg") {
					let jpegData = fs.readFileSync(fileData);
					fileData = jpeg.decode(jpegData, true);
					callback(fileData, fileData.width, fileData.height);
				}
			};
		}

		function isColorSimilar(a, b, color){

			let absDiff = Math.abs(a - b);

			if(typeof a === 'undefined'){
				return false;
			}
			if(typeof b === 'undefined'){
				return false;
			}

			if(a === b){
				return true;
			} else if ( absDiff < tolerance[color] ) {
				return true;
			} else {
				return false;
			}
		}

		function isNumber(n) {
			return !isNaN(parseFloat(n));
		}

		function isPixelBrightnessSimilar(d1, d2){
			let alpha = isColorSimilar(d1.a, d2.a, 'alpha');
			let brightness = isColorSimilar(d1.brightness, d2.brightness, 'minBrightness');
			return brightness && alpha;
		}

		function getBrightness(r,g,b){
			return 0.3*r + 0.59*g + 0.11*b;
		}

		function isRGBSame(d1,d2){
			let red = d1.r === d2.r;
			let green = d1.g === d2.g;
			let blue = d1.b === d2.b;
			return red && green && blue;
		}

		function isRGBSimilar(d1, d2){
			let red = isColorSimilar(d1.r,d2.r,'red');
			let green = isColorSimilar(d1.g,d2.g,'green');
			let blue = isColorSimilar(d1.b,d2.b,'blue');
			let alpha = isColorSimilar(d1.a, d2.a, 'alpha');

			return red && green && blue && alpha;
		}

		function isContrasting(d1, d2){
			return Math.abs(d1.brightness - d2.brightness) > tolerance.maxBrightness;
		}

		function getHue(r,g,b){

			r = r / 255;
			g = g / 255;
			b = b / 255;
			let max = Math.max(r, g, b), min = Math.min(r, g, b);
			let h;
			let d;

			if (max == min){
				h = 0; // achromatic
			} else{
				d = max - min;
				switch(max){
					case r: h = (g - b) / d + (g < b ? 6 : 0); break;
					case g: h = (b - r) / d + 2; break;
					case b: h = (r - g) / d + 4; break;
				}
				h /= 6;
			}

			return h;
		}

		function isAntialiased(sourcePix, data, cacheSet, verticalPos, horizontalPos, width){
			let offset;
			let targetPix;
			let distance = 1;
			let i;
			let j;
			let hasHighContrastSibling = 0;
			let hasSiblingWithDifferentHue = 0;
			let hasEquivilantSibling = 0;

			addHueInfo(sourcePix);

			for (i = distance*-1; i <= distance; i++){
				for (j = distance*-1; j <= distance; j++){

					if(i===0 && j===0){
						// ignore source pixel
					} else {

						offset = ((verticalPos+j)*width + (horizontalPos+i)) * 4;
						targetPix = getPixelInfo(data, offset, cacheSet);

						if(targetPix === null){
							continue;
						}

						addBrightnessInfo(targetPix);
						addHueInfo(targetPix);

						if( isContrasting(sourcePix, targetPix) ){
							hasHighContrastSibling++;
						}

						if( isRGBSame(sourcePix,targetPix) ){
							hasEquivilantSibling++;
						}

						if( Math.abs(targetPix.h - sourcePix.h) > 0.3 ){
							hasSiblingWithDifferentHue++;
						}

						if( hasSiblingWithDifferentHue > 1 || hasHighContrastSibling > 1){
							return true;
						}
					}
				}
			}

			if(hasEquivilantSibling < 2){
				return true;
			}

			return false;
		}

		function errorPixel(px, offset, data1, data2){
			let data = errorPixelTransformer(data1, data2);
			px[offset] = data.r;
			px[offset + 1] = data.g;
			px[offset + 2] = data.b;
			px[offset + 3] = data.a;
		}

		function copyPixel(px, offset, data){
			px[offset] = data.r; //r
			px[offset + 1] = data.g; //g
			px[offset + 2] = data.b; //b
			px[offset + 3] = data.a * pixelTransparency; //a
		}

		function copyGrayScalePixel(px, offset, data){
			px[offset] = data.brightness; //r
			px[offset + 1] = data.brightness; //g
			px[offset + 2] = data.brightness; //b
			px[offset + 3] = data.a * pixelTransparency; //a
		}

		function getPixelInfo(data, offset, cacheSet){
			let r;
			let g;
			let b;
			let d;
			let a;

			r = data[offset];

			if(typeof r !== 'undefined'){
				g = data[offset+1];
				b = data[offset+2];
				a = data[offset+3];
				d = {
					r: r,
					g: g,
					b: b,
					a: a
				};

				return d;
			} else {
				return null;
			}
		}

		function addBrightnessInfo(data){
			data.brightness = getBrightness(data.r,data.g,data.b); // 'corrected' lightness
		}

		function addHueInfo(data){
			data.h = getHue(data.r,data.g,data.b);
		}

		function analyseImages(img1, img2, width, height){

			let data1 = img1.data;
			let data2 = img2.data;

      //TODO
      let imgd = new PNG({
          width: img1.width,
          height: img1.height,
          deflateChunkSize: img1.deflateChunkSize,
          deflateLevel: img1.deflateLevel,
          deflateStrategy: img1.deflateStrategy,
        });
			let targetPix = imgd.data;

			let mismatchCount = 0;

			let time = Date.now();

			let skip;

      let currentRectangle = null;
      let rectagnlesIdx = 0;

			if(!!largeImageThreshold && ignoreAntialiasing && (width > largeImageThreshold || height > largeImageThreshold)){
				skip = 6;
			}

			loop(height, width, function(verticalPos, horizontalPos){

				let offset = (verticalPos*width + horizontalPos) * 4;

				if(skip){ // only skip if the image isn't small
					if(verticalPos % skip === 0 || horizontalPos % skip === 0){

						copyPixel(targetPix, offset, {
							r: 0,
							b: 0,
							g: 0,
							a: 0
						});

						return;
					}
				}

				let pixel1 = getPixelInfo(data1, offset, 1);
				let pixel2 = getPixelInfo(data2, offset, 2);

				if(pixel1 === null || pixel2 === null){
					return;
				}

        if (ignoreRectangles) {
          for(rectagnlesIdx = 0; rectagnlesIdx < ignoreRectangles.length; rectagnlesIdx++) {
            currentRectangle = ignoreRectangles[rectagnlesIdx];
            //console.log(currentRectangle, verticalPos, horizontalPos);
            if (
              (verticalPos >= currentRectangle[1]) &&
              (verticalPos < currentRectangle[1] + currentRectangle[3]) &&
              (horizontalPos >= currentRectangle[0]) &&
              (horizontalPos < currentRectangle[0] + currentRectangle[2])
            ) {
						  copyGrayScalePixel(targetPix, offset, pixel2);
					    //copyPixel(targetPix, offset, pixel1, pixel2);
              return;
            }
          }
        }

				if (ignoreColors){

					addBrightnessInfo(pixel1);
					addBrightnessInfo(pixel2);

					if( isPixelBrightnessSimilar(pixel1, pixel2) ){
						copyGrayScalePixel(targetPix, offset, pixel2);
					} else {
						errorPixel(targetPix, offset, pixel1, pixel2);
						mismatchCount++;
					}
					return;
				}

				if( isRGBSimilar(pixel1, pixel2) ){
					copyPixel(targetPix, offset, pixel1);

				} else if( ignoreAntialiasing && (
						addBrightnessInfo(pixel1), // jit pixel info augmentation looks a little weird, sorry.
						addBrightnessInfo(pixel2),
						isAntialiased(pixel1, data1, 1, verticalPos, horizontalPos, width) ||
						isAntialiased(pixel2, data2, 2, verticalPos, horizontalPos, width)
					)){

					if( isPixelBrightnessSimilar(pixel1, pixel2) ){
						copyGrayScalePixel(targetPix, offset, pixel2);
					} else {
						errorPixel(targetPix, offset, pixel1, pixel2);
						mismatchCount++;
					}
				} else {
					errorPixel(targetPix, offset, pixel1, pixel2);
					mismatchCount++;
				}

			});

			data.rawMisMatchPercentage = (mismatchCount / (height*width) * 100);
			data.misMatchPercentage = data.rawMisMatchPercentage.toFixed(2);
			data.analysisTime = Date.now() - time;

			data.getDiffImage = function(text){
        return imgd;
			};

			data.getDiffImageAsJPEG = function(quality) {
				return jpeg.encode({
				  data: targetPix,
				  width: img1.width,
				  height: img1.height
				}, quality !== undefined ? quality : 50).data;
			};
		}

		function compare(one, two){

			function onceWeHaveBoth(img){
				let width;
				let height;

        images.push(img);
				if(images.length === 2){
					width = images[0].width > images[1].width ? images[0].width : images[1].width;
					height = images[0].height > images[1].height ? images[0].height : images[1].height;

					if( (images[0].width === images[1].width) && (images[0].height === images[1].height) ){
						data.isSameDimensions = true;
					} else {
						data.isSameDimensions = false;
					}

					data.dimensionDifference = { width: images[0].width - images[1].width, height: images[0].height - images[1].height };

          //lksv: normalization removed
					analyseImages( images[0], images[1], width, height);

					triggerDataUpdate();
				}
			}

			images = [];
			loadImageData(one, onceWeHaveBoth);
			loadImageData(two, onceWeHaveBoth);
		}

		function getCompareApi(param){

			let secondFileData,
				hasMethod = typeof param === 'function';

			if( !hasMethod ){
				// assume it's file data
				secondFileData = param;
			}

			let self = {
				ignoreNothing: function(){

					tolerance.red = 0;
					tolerance.green = 0;
					tolerance.blue = 0;
					tolerance.alpha = 0;
					tolerance.minBrightness = 0;
					tolerance.maxBrightness = 255;

					ignoreAntialiasing = false;
					ignoreColors = false;

					if(hasMethod) { param(); }
					return self;
				},
				ignoreAntialiasing: function(){

					tolerance.red = 32;
					tolerance.green = 32;
					tolerance.blue = 32;
					tolerance.alpha = 32;
					tolerance.minBrightness = 64;
					tolerance.maxBrightness = 96;

					ignoreAntialiasing = true;
					ignoreColors = false;

					if(hasMethod) { param(); }
					return self;
				},
				ignoreColors: function(){

					tolerance.alpha = 16;
					tolerance.minBrightness = 16;
					tolerance.maxBrightness = 240;

					ignoreAntialiasing = false;
					ignoreColors = true;

					if(hasMethod) { param(); }
					return self;
				},
        //array of rectangles, each rectangle is defined as (x, y, width. height)
        //e.g. [[325, 170, 100, 40]]
        ignoreRectangles: function(rectangles) {
          ignoreRectangles = rectangles;
          return self;
        },
				repaint: function(){
					if(hasMethod) { param(); }
					return self;
				},
				onComplete: function( callback ){

					updateCallbackArray.push(callback);

					let wrapper = function(){
						compare(fileData, secondFileData);
					};

					wrapper();

					return getCompareApi(wrapper);
				}
			};

			return self;
		}

		return {
			onComplete: function( callback ){
				updateCallbackArray.push(callback);
				loadImageData(fileData, function(imageData, width, height){
					parseImage(imageData.data, width, height);
				});
			},
			compareTo: function(secondFileData){
				return getCompareApi(secondFileData);
			}
		};

	};

	_this['resemble'].outputSettings = function(options){
		let key;
		let undefined;

		if(options.errorColor){
			for (key in options.errorColor) {
				errorPixelColor[key] = options.errorColor[key] === undefined ? errorPixelColor[key] : options.errorColor[key];
			}
		}

		if(options.errorType && errorPixelTransform[options.errorType] ){
			errorPixelTransformer = errorPixelTransform[options.errorType];
		}

		pixelTransparency = options.transparency || pixelTransparency;

		if (options.largeImageThreshold !== undefined) {
			largeImageThreshold = options.largeImageThreshold;
		}

		return this;
	};

module.exports = _this['resemble']
